feat: cleanup FSM code to contain real abstractions instead of directly applying low-level functions.

bollu opened the pull request.

bollu committed:
chore: delete dead code

bollu committed:
chore: try to add more abstractions

bollu committed:
chore: prove eval eq eval'

bollu committed:
chore: get main thm through

alexkeizer commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+variable {σ β σ' β' : Type} {γ : β → Type}
+
+/-- An `arity` indexed product of Booleans. -/
+def BoolProd (arity : Type) : Type := arity → Bool
```
If we're going to refactor this, I think I'd prefer to just pay the cost once, change `arity` to a Nat, and use `BitVec w`

alexkeizer commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+  fun a => (x a).eval y
+
+/-- A 'arity' indexed product of `Bitstream`s. -/
+def BitStreamProd (arity : Type) : Type := arity → BitStream
```
```suggestion
def BitStreams (arity : Nat) : Type := Fin arity → BitStream
```

alexkeizer replied:
Although I fully admit I'm being self-contradictory, given my previous crusade against semantically similar definitions having a <= 1 edit distance, somehow singular vs plural feels like a safe exception to that rule.

Otherwise, I'd be more comfortable even calling it a `BitStreamVec`(tor) --- although that is tempting confusion with `BitVec`tors (aagh, why is naming so hard). In my mind, `Prod` means binary product, an n-ary product is a Vector

alexkeizer commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+  fun i => (x i).head
+
+def BitStreamProd.tail (x : BitStreamProd arity) : BitStreamProd arity :=
+  fun a => (x a).tail
```
These could use some docstring explicitly explaining you're doing `head`/`tail` on the constituent bitstreams! I was a bit confused by the arity being the same on both sides at first.

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
-  [ dec_eq : DecidableEq α ]
+  The arity of the (finite) type `σ` (for `s`tate) determines how many bits the
+  internal carry state of this FSM has -/
+  ( σ  : Type ) -- Why is σ also not an index?
```
```suggestion
  internal carry state of this FSM has. -/
  ( σ  : Type ) -- Why is σ also not an index?
```

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+
+Note that `p.σ` is assumed to be a finite type, so `p.State` is morally
+a finite bitvector whose width is given by the arity of `p.σ` -/
+abbrev State (p : FSM arity): Type := BoolProd p.σ
```
```suggestion
a finite bitvector whose width is given by the arity of `p.σ`. -/
abbrev State (p : FSM arity): Type := BoolProd p.σ
```

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+  (p.currentOutCircuit).eval (state.appendInput input)
+
+
+end FSM
```
```suggestion

end FSM
```

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+    outputStreamAux p (p.nextState s₀ (inputStream.head)) inputStream.tail n := by rfl
+
+
+def StateStream (p : FSM arity) := ℕ → p.State
```
```suggestion

def StateStream (p : FSM arity) := ℕ → p.State
```

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -11,107 +11,191 @@ import SSA.Experimental.Bits.Fast.Circuit
+def StateStream (p : FSM arity) := ℕ → p.State
+
+/-- `p.carryStream inputStream` computes the stream of carries, from the stream of inputs -/
+def carryStream (inputStream : BitStreamProd arity) : p.StateStream := fun n =>
```
```suggestion
/-- `p.carryStream inputStream` computes the stream of carries, from the stream of inputs. -/
def carryStream (inputStream : BitStreamProd arity) : p.StateStream := fun n =>
```

alexkeizer committed:
WIP: FSM abstractions

alexkeizer committed:
appendVector'

luisacicolini committed:
tmp

alexkeizer committed:
WIP: CircuitProd helpers

alexkeizer committed:
Merge branch 'fsm-abstractions' of github.com:opencompl/lean-mlir into fsm-abstractions

alexkeizer committed:
fix: red squigly

Co-authored-by: Siddharth <siddu.druid@gmail.com>

alexkeizer committed:
WIP: introduce more `BitVec` abstractions

Co-authored-by: Siddharth <siddu.druid@gmail.com>

alexkeizer committed:
refactor: move refactored defs to a `New` suffixed copy of FiniteStateMachine

Co-authored-by: Siddharth <siddu.druid@gmail.com>

alexkeizer committed:
final final attempt

Let's not continue this

alexkeizer committed:
nuke bitvectors

Co-authored-by: Siddharth <siddu.druid@gmail.com>
Co-authored-by: Luisa Cicolini <luisacicolini@users.noreply.github.com>

luisacicolini committed:
cleaned bitwise op section a bit

luisacicolini committed:
name fixes

luisacicolini committed:
pairing with sid: bitwiseAnd, its eval and support stuff

bollu committed:
chore: bring back eval and eval'.

This lets us use the low level corecursion definition using eval',
and the theorem `eval_eq_eval'` to go between the higher level
and lower level versions.

bollu committed:
chore: prove bitwiseAnd

We circumvent the prior proof of high technology with map and evalMap,
instead going for directly defining the FSM for bitwise and.

This method has the distinct advantage of actually being able to be written
with the current broken defs.

This also forces us to add quite a bit of lemmata, which is always nice.

bollu committed:
chore: make eval_bitwiseAnd not depend on sorry anymore

bollu committed:
feat: faff about with subsingleton, hoping for a cleanup, but none was achieved

luisacicolini committed:
bitwise operators

bollu committed:
chore: document chris proof

tobiasgrosser commented on SSA/Experimental/Bits/Fast/FiniteStateMachine.lean:
```diff
@@ -463,10 +469,21 @@ theorem carry_decr (x : Unit → BitStream) : ∀ (n : ℕ), decr.carry x (n+1)
+If the set of states R is closed under the inverse image of next state,
+and if every state in the set R produces a true next output,
+and furthermore, if our initial state is not R,
+the the value at the initial state is zero, and furthermore, the state will not enter into R in the next state
```
```suggestion
the value at the initial state is zero, and furthermore, the state will not enter into R in the next state
```

luisacicolini committed:
bitwise fsm cleaned

luisacicolini committed:
bitwise fsm cleaned

luisacicolini committed:
either enlightnening or nonsense

luisacicolini committed:
i am giving up on this predicate

luisacicolini committed:
peiring with sid, binary predicates

luisacicolini committed:
satisfactory theorem decl but too fried to prove it now

luisacicolini committed:
progress with eval_and

luisacicolini committed:
it was easier than i tought

luisacicolini committed:
lost again

tobiasgrosser commented:
This has now been superseeded, I guess.
